test/hotspot/jtreg/serviceability/sa/ClhsdbJstackWithConcurrentLock.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8324066
+ * @summary Test the clhsdb 'jstack -l' command for printing concurrent lock information
+ * @requires vm.hasSA
+ * @library /test/lib
+ * @run main/othervm ClhsdbJstackWithConcurrentLock
+ */
+
+import java.util.List;
+import jdk.test.lib.apps.LingeredApp;
+import jtreg.SkippedException;
+
+public class ClhsdbJstackWithConcurrentLock {
+
+    public static void main(String[] args) throws Exception {
+        System.out.println("Starting the ClhsdbJstackWithConcurrentLock test");
+
+        LingeredApp theApp = null;
+        try {
+            ClhsdbLauncher test = new ClhsdbLauncher();
+
+            theApp = new LingeredAppWithConcurrentLock();
+            // Use a small heap so the scan is quick.
+            LingeredApp.startApp(theApp, "-Xmx4m");
+            System.out.println("Started LingeredApp with pid " + theApp.getPid());
+
+            // Run the 'jstack -l' command to get the stack and have java.util.concurrent
+            // lock information included.
+            List<String> cmds = List.of("jstack -l");
+            String jstackOutput = test.run(theApp.getPid(), cmds, null, null);
+
+            // We are looking for:
+            //   Locked ownable synchronizers:
+            //    - <0x00000000ffc2ed70>, (a java/util/concurrent/locks/ReentrantLock$NonfairSync)
+            // We want to fetch the address from this line.
+            String key = ", (a java/util/concurrent/locks/ReentrantLock$NonfairSync)";
+            String[] lines = jstackOutput.split("\\R");
+            String addressString = null;
+            for (String line : lines) {
+                if (line.contains(key)) {
+                    String[] words = line.split(key + "|[, ]");
```

Comment:
String.split() expect regexp as an argument
Not sure how this code works (without escaping '$' and parentheses).
I think it would be clearer to use standard regexp pattern/matcher, something like
`
Pattern pattern = Pattern.compile("\\s+- <0x([0-9a-fA-F]+)>, \\(a java/util/concurrent/locks/ReentrantLock\\$NonfairSync\\)");
for (String line: lines) {
    Matcher matcher = pattern.matcher(line);
    if (matcher.find()) {
        addressString = matcher.group(1);
        break;
    }
`}

Reply:
I think `key` doesn't really matter when doing the split. `key` is mainly for finding the right line. The split is for tokenizing that line into words, and the only parts that matters for that are ' ' and the ',' being used as word delimiters. I think `key` should just be removed from the regex. I think the source of the bug is code copied from ClhsdbInspect.java:

                        // Escape the token "Method*=" because the split method uses
                        // a regex, not just a straight String.
                        String escapedKey = key.replace("*","\\*");
                        String[] words = line.split(escapedKey+"|[ ]");

In this case key is `Method*=`. The code works and the escaping and searching for `key` is necessary because we are looking for something like `Method*=<0x00000000ffc2ed70>`, but I think this could have been simplified by including `=` in the split pattern rather than all of `escapedKey`.

Reply:
Well, I see now.
In this case `"[, ]"` separator should work fine.
I still think that use of Matcher is much clear - regexp defines what you are looking for and extract the value you need. Also there is no need to split `jstackOutput` into lines - the whole jstack output can be passed to `Pattern.matcher`.
But I don't force you if you prefer to extract the address manually.